### PR TITLE
feat(engine): checkpoint in memory 

### PR DESCRIPTION
**Pull request by soodoshll**
- with its python interface
- single threading (can be improved)

### CODE REVIEW

File: src/engine/archive.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        std::map<const Flow *, FlowArchive> flowArchive;
+        std::map<const TrafficLight *, TrafficLightArchive> trafficLightArchive;
+        size_t step;
+        size_t activeVehicleCount;
+        std::mt19937 rnd;
+
+        Vehicle * getNewPointer(const Vehicle * old);
+        void archiveDrivable(const Drivable *drivable, DrivableArchive &archive);
+        void archiveFlow(const Flow *flow, FlowArchive &archive);
+        void archiveTrafficLight(const TrafficLight *light, TrafficLightArchive &archive);
```
[the remark this answers is not in the document]
Comment (soodoshll) [reply src/engine/archive.h]: These functions call the `getNewPointer` function which changes old vehicle pointers to new (according to duplicated vehiclePool, a `std::map`)

File: src/engine/archive.cpp [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        archive.cnt = flow->cnt;
+    }
+
+    void Archive::archiveTrafficLight(const TrafficLight *light, Archive::TrafficLightArchive &archive) {
+        archive.curPhaseIndex = light->curPhaseIndex;
+        archive.remainDuration = light->remainDuration;
+    }
+
+
+    void Archive::resume(Engine &engine) const{
```
[the remark this answers is not in the document]
Comment (soodoshll) [reply src/engine/archive.cpp]: The preferred API is Engine::load(const Archive &archive) . In fact, I think Archive::resume should not be public to users.

File: src/engine/archive.cpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        for (const auto &pair : vehiclePool) {
+            const auto &vehicle = pair.second.first;
+            size_t threadIndex = pair.second.second;
+            engine.threadVehiclePool[threadIndex].emplace(vehicle);
+        }
+        for (auto drivable : engine.roadnet.getDrivables()) {
+            const auto &archive = drivableArchive.find(drivable)->second;
+            drivable->vehicles = archive.vehicles;
+            if (drivable->isLane()) {
+                Lane * lane = dynamic_cast<Lane *>(drivable);
```
Comment (Hzfengsy) [inline src/engine/archive.cpp]: code style `Lane *lane`

File: src/engine/archive.cpp [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        engine.vehiclePool = vehiclePool;
+        engine.rnd = rnd;
+        for (auto &threadVeh : engine.threadVehiclePool)
+            threadVeh.clear();
+        for (const auto &pair : vehiclePool) {
+            const auto &vehicle = pair.second.first;
+            size_t threadIndex = pair.second.second;
+            engine.threadVehiclePool[threadIndex].emplace(vehicle);
+        }
+        for (auto drivable : engine.roadnet.getDrivables()) {
```
Comment (Hzfengsy) [inline src/engine/archive.cpp]: I'm not sure why we don't need to use reference here.

File: src/engine/archive.cpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            archive.vehicles.emplace_back(getNewPointer(vehicle));
+        if (drivable->isLane()) {
+            const Lane * lane = dynamic_cast<const Lane *>(drivable);
+            for (const auto &vehicle : lane->getWaitingBuffer()) {
+                archive.waitingBuffer.emplace_back(getNewPointer(vehicle));
+            }
+        }
+    }
+
+    void Archive::archiveFlow(const Flow *flow, Archive::FlowArchive &archive) {
```
Comment (Hzfengsy) [inline src/engine/archive.cpp]: I prefer to rename `archive` to `flowArchive`. Name `archive` is confusing and I often think it is a variable of `class Archive`

File: src/engine/archive.cpp [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    Vehicle *Archive::getNewPointer(const Vehicle *old) {
+        if (!old) return nullptr;
+        int priority = old->getPriority();
+        assert(vehiclePool.count(priority));
+        return vehiclePool[priority].first;
+    }
+
+    void Archive::archiveDrivable(const Drivable *drivable, Archive::DrivableArchive &archive) {
```
Comment (Hzfengsy) [inline src/engine/archive.cpp]: The same as below

File: src/engine/archive.cpp [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        }
+    }
+
+    void Archive::archiveFlow(const Flow *flow, Archive::FlowArchive &archive) {
+        archive.currentTime = flow->currentTime;
+        archive.nowTime = flow->nowTime;
+        archive.cnt = flow->cnt;
+    }
+
+    void Archive::archiveTrafficLight(const TrafficLight *light, Archive::TrafficLightArchive &archive) {
```
Comment (Hzfengsy) [inline src/engine/archive.cpp]: The same as above

File: src/engine/archive.h [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+        std::map<int, std::pair<Vehicle *, int>> vehiclePool;
+        std::map<const Drivable *, DrivableArchive> drivableArchive;
+        std::map<const Flow *, FlowArchive> flowArchive;
+        std::map<const TrafficLight *, TrafficLightArchive> trafficLightArchive;
+        size_t step;
+        size_t activeVehicleCount;
+        std::mt19937 rnd;
+
+        Vehicle * getNewPointer(const Vehicle * old);
```
Comment (Hzfengsy) [inline src/engine/archive.h]: code style `Vehicle *getNewPointer`

File: src/engine/archive.cpp [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+namespace CityFlow {
+
+    Archive::Archive(const Engine &engine)
+    : step(engine.step), activeVehicleCount(engine.activeVehicleCount), rnd(engine.rnd) {
+        // copy the vehicle Pool
+
+        for (const auto &veh : engine.vehiclePool) {
+            const Vehicle *oldVehicle = veh.second.first;
+            Vehicle * newVehicle = new Vehicle(*oldVehicle);
+            vehiclePool.emplace(oldVehicle->getPriority(), std::make_pair(newVehicle, veh.second.second));
```
Comment (Hzfengsy) [inline src/engine/archive.cpp]: Does copying during construction mean one archive can be loaded only once?

Comment (soodoshll) [reply src/engine/archive.cpp]: > Does copying during construction mean one archive can be loaded only once?

This is a mistake caused by my neglect. I think there are two choices:

-  Set a flag in each Archive object, which stands for whether this archive has been 'used'. Thus we can avoid twice loading.  The user need to create another copy manually.

- After an archive is loaded, construct a new archive immediately and replace the former one.

The first way is better in performance but easy to cause mistakes while the latter is easy to use but may cause waste.

File: src/engine/archive.cpp (hunk 6 above)
Comment (zhc134) [inline src/engine/archive.cpp]: HistoryRecord should also be archived

Comment (zhc134) [reply src/engine/archive.cpp]: also, please change `historyRecord` to `HistoryRecord`

File: src/test_archive.cpp [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+#include <functional>
+#include <chrono>
+
+using namespace CityFlow;
+namespace bpo = boost::program_options;
+constexpr int interval = 100;
+
+double getDuration(const clock_t &a, const clock_t &b) {
+    return ((double)(a - b)) / CLOCKS_PER_SEC;
+}
```
Comment (Hzfengsy) [inline src/test_archive.cpp]: This timer method calculates CPU time, which means if we use multi-threading, the result will be several times larger than real time.

File: src/test_archive.cpp [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    std::string configFile;
+    int threadNum;
+
+    bpo::options_description opts("all options");
+    bpo::variables_map vm;
+
+    opts.add_options()
+            ("configFile", bpo::value<std::string>(&configFile)->required(), "config file")
+            ("threadNum", bpo::value<int>(&threadNum)->default_value(1), "number of threads")
+            ("help", "Simulator Parameters");
```
Comment (Hzfengsy) [inline src/test_archive.cpp]: It is only a test rather than a "simulator".

File: src/vehicle/vehicle.cpp [hunk 12]
```diff
@@ -397,4 +408,8 @@ namespace CityFlow {
         laneChangeInfo.partner = nullptr;
         laneChangeInfo.offset = 0;
     }
+
```
Comment (Hzfengsy) [inline src/vehicle/vehicle.cpp]: Don't leave too many blank lines.

File: src/test_archive.cpp [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    Engine engine(dataDir + "/config/" + configFile, (unsigned int)threadNum);
+
+    std::vector<std::function<bool(Engine &engine)>> testFunctions{test0, test1, test2, test3};
+    std::vector<std::string> captions{
+        "Single save, single load",
+        "Single save, multiple loads",
+        "Multiple saves, multiple loads",
+        "Performance test"
+        };
+    for (int i = 0 ; i < interval ; ++i) engine.nextStep();
```
Comment (Hzfengsy) [inline src/test_archive.cpp]: Why it names "interval"? And please change `int` to `size_t`

File: src/test_archive.cpp [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+bool test0(Engine &engine)
+{
+    auto before = engine.getLaneVehicleCount();
+    Archive archive(engine);
+    for (int i = 0 ; i < interval ; ++i) engine.nextStep();
+    auto after1 = engine.getLaneVehicleCount();
+    engine.load(archive);
+    for (int i = 0 ; i < interval ; ++i) engine.nextStep();
+    auto after2 = engine.getLaneVehicleCount();
+    return after1 == after2;
```
Comment (Hzfengsy) [inline src/test_archive.cpp]: I don't think it is enough that we only check `LaneVehicleCount`.

Comment (Hzfengsy) [reply src/test_archive.cpp]: BTW, should we add some more APIs, like `getAvarageSpeed`, `getAvarageTravelTime` or whatever? I think it is useful in both testing and RL training. @zhc134

File: src/test_archive.cpp [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+    return after1 == after2 && after2 == after3;
+}
+
+bool test2(Engine &engine)
+{
+    /*
+     *  step     0   100 200 300 400
+     *  archive  0   1   2   3   4
+     */
+    constexpr int intervalNum = 5;
```
Comment (Hzfengsy) [inline src/test_archive.cpp]: also, why "interval"?

File: src/test_archive.cpp [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+                    return false;
+            }
+        }
+    }
+    return true;
+}
+
+bool test3(Engine &engine)
+{
+    constexpr int repeat_times = 10;
```
Comment (Hzfengsy) [inline src/test_archive.cpp]: code style `repeatTimes` or change it to `repeats`

File: src/vehicle/lanechange.cpp [hunk 17]
```diff
@@ -1,35 +1,43 @@
...
 #include "vehicle/vehicle.h"
 #include "engine/engine.h"
 
 namespace CityFlow{
+    LaneChange::LaneChange(Vehicle * vehicle, const LaneChange &other)
+        : lastDir(other.lastDir), signalRecv(other.signalRecv),
+          vehicle(vehicle), targetLeader(other.targetLeader), targetFollower(other.targetFollower), // useless in archive
+          leaderGap(other.leaderGap), followerGap(other.followerGap), waitingTime(other.waitingTime),
+          changing(other.changing), lastChangeTime(other.lastChangeTime)
+    {
```
Comment (zhc134) [inline src/vehicle/lanechange.cpp]: code style "{" should not start new line

File: CMakeLists.txt [hunk 18]
```diff
@@ -52,8 +52,9 @@ add_subdirectory(src)
 find_package(Boost 1.50 REQUIRED COMPONENTS
 	program_options
 )
+
 add_executable(${PROJECT_NAME}_test src/test.cpp)
-target_link_libraries(${PROJECT_NAME}_test PUBLIC ${PROJECT_LIB_NAME} ${Boost_LIBRARIES})
+target_link_libraries( ${PROJECT_NAME}_test PUBLIC ${PROJECT_LIB_NAME} ${Boost_LIBRARIES})
```
Comment (zhc134) [inline CMakeLists.txt]: remove space

File: src/simulator.cpp [hunk 19]
```diff
@@ -24,8 +25,11 @@ PYBIND11_MODULE(cityflow, m) {
         .def("get_current_time", &CityFlow::Engine::getCurrentTime)
         .def("set_tl_phase", &CityFlow::Engine::setTrafficLightPhase)
         .def("push_vehicle", (void (CityFlow::Engine::*)(const std::map<std::string, double>&, const std::vector<std::string>&)) &CityFlow::Engine::pushVehicle)
-        .def("reset", &CityFlow::Engine::reset);
-        
+        .def("reset", &CityFlow::Engine::reset)
+        .def("load", &CityFlow::Engine::load);
```
Comment (zhc134) [inline src/simulator.cpp]: add snapshot() api

**Pull request merged**